File: LeetCode/src/dp/ClimbingStairs.java
package dp;

public class ClimbingStairs {
    public int climbStairs(int n) {
        if (n <= 2) {
            return n;
        }

        int[] dp = new int[n + 1];
        dp[1] = 1;
        dp[2] = 2;

        for (int i = 3; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }

        return dp[n];
    }

//    public int climbStairs(int n) {
//        if (n <= 2) {
//            return n;
//        }
//
//        int dp1 = 2;
//        int dp2 = 1;
//
//        for (int i = 3; i <= n; i++) {
//            int dp = dp1 + dp2;
//            dp2= dp1;
//            dp1 = dp;
//        }
//
//        return dp1;
//    }

    public static void main(String[] args) {
        ClimbingStairs solution = new ClimbingStairs();
        int n = 4;
        System.out.println(solution.climbStairs(n)); // Output: 5
    }
}

File: LeetCode_Python/dp/ClimbingStairs.py
class ClimbingStairs:
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n

        first = 1
        second = 2

        for i in range(3, n + 1):
            third = first + second
            first = second
            second = third

        return second

    # Time Complexity: O(n)
    # Space Complexity: O(1)

    def climbStairsByDPArray(self, n: int) -> int:
        if n <= 2:
            return n

        dp = [0] * (n + 1)
        dp[1] = 1
        dp[2] = 2

        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]

        return dp[n]

    # Time Complexity: O(n)
    # Space Complexity: O(n)


if __name__ == "__main__":
    solution = ClimbingStairs()

    print("Expected Output : 2")
    print("Actual Output : ", solution.climbStairs(2))
    print("Actual Output : ", solution.climbStairsByDPArray(2))

    print("Expected Output : 3")
    print("Actual Output : ", solution.climbStairs(3))
    print("Actual Output : ", solution.climbStairsByDPArray(3))
